Add Best Practice: Using unique output files and directories (#35909)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc

=== Tags
`<<tags_reference.adoc#tag:tasks,#tasks>>`, `<<tags_reference.adoc#tag:inputs-and-outputs,#inputs-and-outputs>>`, `<<tags_reference.adoc#tag:configurations,#configurations>>`

[[use_unique_output_files_and_directories]]
== Use unique output files and directories

Overlapping output files or directories cause tasks to rerun unnecessarily and waste work.

=== Explanation

Gradle tracks all output files and directories declared by tasks to decide whether a task needs to be rerun.

> For example, if the contents of a task’s output directory change after its last execution, Gradle will rerun that task.

Reviewer: What do you think of:
```suggestion
For example, if the contents of a task’s output directory change between builds, Gradle will treat the task as out-of-date and execute it again.
```

Ensuring that each task uses its own unique output files and directories, both within a project and across the entire build, prevents unnecessary work.

=== Example

==== Don't Do This

====
include::sample[dir="snippets/bestPractices/uniqueOutputs-avoid/kotlin",files="build.gradle.kts[tags=avoid-this]"]
include::sample[dir="snippets/bestPractices/uniqueOutputs-avoid/groovy",files="build.gradle[tags=avoid-this]"]
====

<1> *Write to a file in the output directory*: This task produces a single file in the `outputDirectory`, named based on the `type` input.
<2> *Read a specific file in the input directory*: This task only needs to read a single `a.txt` file in the input directory.
<3> *Set output directory*: Sets `outputDirectory` to a subdirectory in `buildDirectory`.
<4> *Set output directory*: Same as above, using the same shared `greetings` directory.
<5> *Wire `greeterA` to consumer*: Makes sure that `greeterA` runs and produces the output directory before it is used by `consumer`.

With this setup, if you run the `consumer` task, then `greeterB`, the `consumer` task will be invalidated.

> The next time `consumer` is run it will *not* be `UP-TO-DATE` and will have to run again despite not using the output from `greeterB`.

Reviewer: ```suggestion
The next time you run `consumer`, it will *not* be `UP-TO-DATE` and will execute again, even though it does not use the output from `greeterB`.
```

This happens because `greeterB` changes the contents of the shared output directory `greetings`, which is an output of `greeterA` that `consumer` depends on (despite `consumer` only actually using the unchanged `a.txt` file in that directory).

==== Do This Instead

> To avoid issues, avoid using shared task output directories and files.

Reviewer: ```suggestion
To avoid issues, avoid shared output directories or files across multiple tasks.
```

Instead, tasks should only declare the exact outputs and consume the exact inputs that they actually produce and consume.

The _simplest_ change to make here is to use distinct output directories for each `GreetingTask`.

> This alone is sufficient to fix the problem.

Reviewer: ```suggestion
This change alone is enough to prevent `consumer` from being invalidated when `greeterB` runs:
```

====
include::sample[dir="snippets/bestPractices/uniqueOutputs-simple-do/kotlin",files="build.gradle.kts[tags=simple-do-this]"]
include::sample[dir="snippets/bestPractices/uniqueOutputs-simple-do/groovy",files="build.gradle[tags=simple-do-this]"]
====

<1> *Set unique output directories*: Each `GreetingTask` is assigned its own unique output directory based on the `type` input.

Now when running `consumer` task, then `greeterB`, then `consumer` task remains `UP-TO-DATE` as Gradle knows that it is not using the output from `greeterB`, since `greeterA` and `greeterB` write to distinct output directories.

However, a more _complete and idiomatic_ approach realizes that:

> 1. Tasks that produce single output files should make this clear from the type of their `@Output` properties.

Reviewer: This isn't quite clear to me. What about something like:
```suggestion
1. Tasks that produce a single output file should express this clearly by using a single-file type `@Output` property.
```

> 1. Tasks that only consume single input files should make this clear from the type of their `@Input` properties.

Reviewer: ```suggestion
2. Tasks that consume a single input file should likewise use a single-file `@Input` property.
```
